File: backend/internal/model/challenge.go
package model

import (
	"time"
)

// Challenge 挑战模型
type Challenge struct {
	ID              string     `gorm:"type:char(36);primaryKey" json:"id"`
	RouteID         string     `gorm:"type:char(36);not null;index:idx_route" json:"route_id"`
	ChallengerID    string     `gorm:"type:char(36);not null;index:idx_challenger" json:"challenger_id"`
	ChallengerRunID *string    `gorm:"type:char(36)" json:"challenger_run_id,omitempty"`
	InviteeID       *string    `gorm:"type:char(36);index:idx_invitee" json:"invitee_id,omitempty"`
	// TargetRunID 被挑战的目标跑步记录ID（异步挑战：B挑战A的历史记录）
	TargetRunID     *string    `gorm:"type:char(36);index:idx_target_run" json:"target_run_id,omitempty"`
	GhostMode       string     `gorm:"type:varchar(20);default:'real_replay'" json:"ghost_mode"`
	GoalMetric      *string    `gorm:"type:varchar(20)" json:"goal_metric,omitempty"`
	Status          string     `gorm:"type:varchar(20);default:'pending';index:idx_status" json:"status"`
	ChallengerResult *string   `gorm:"type:json" json:"challenger_result,omitempty"`
	InviteeResult   *string    `gorm:"type:json" json:"invitee_result,omitempty"`
	WinnerID        *string    `gorm:"type:char(36)" json:"winner_id,omitempty"`
	CreatedAt       time.Time  `gorm:"type:datetime(3)" json:"created_at"`
	AcceptedAt      *time.Time `gorm:"type:datetime(3)" json:"accepted_at,omitempty"`
	StartedAt       *time.Time `gorm:"type:datetime(3)" json:"started_at,omitempty"`
	CompletedAt     *time.Time `gorm:"type:datetime(3)" json:"completed_at,omitempty"`
	ExpiresAt       *time.Time `gorm:"type:datetime(3);index:idx_expires" json:"expires_at,omitempty"`

	// 关联
	Route       Route        `gorm:"foreignKey:RouteID" json:"route,omitempty"`
	Challenger  User         `gorm:"foreignKey:ChallengerID" json:"challenger,omitempty"`
	Invitee     *User        `gorm:"foreignKey:InviteeID" json:"invitee,omitempty"`
	Winner      *User        `gorm:"foreignKey:WinnerID" json:"winner,omitempty"`
	TargetRun   *Run         `gorm:"foreignKey:TargetRunID" json:"target_run,omitempty"`
	Comparison  *Comparison  `gorm:"foreignKey:ChallengeID" json:"comparison,omitempty"`
}

func (Challenge) TableName() string {
	return "challenges"
}

// Comparison 对比报告模型
type Comparison struct {
	ID           string    `gorm:"type:char(36);primaryKey" json:"id"`
	ChallengeID  *string   `gorm:"type:char(36);uniqueIndex:idx_challenge" json:"challenge_id,omitempty"`
	RunAID       string    `gorm:"type:char(36);not null;index:idx_run_a" json:"run_a_id"`
	RunBID       string    `gorm:"type:char(36);not null;index:idx_run_b" json:"run_b_id"`
	OverallDiff  string    `gorm:"type:json;not null" json:"overall_diff"`
	SplitsJSON   *string   `gorm:"type:json" json:"splits_json,omitempty"`
	DiagnosisJSON *string  `gorm:"type:json" json:"diagnosis_json,omitempty"`
	CreatedAt    time.Time `gorm:"type:datetime(3)" json:"created_at"`

	Challenge *Challenge `gorm:"foreignKey:ChallengeID" json:"challenge,omitempty"`
	RunA      Run        `gorm:"foreignKey:RunAID" json:"run_a,omitempty"`
	RunB      Run        `gorm:"foreignKey:RunBID" json:"run_b,omitempty"`
}

func (Comparison) TableName() string {
	return "comparisons"
}

// Friendship 好友关系模型
type Friendship struct {
	ID        string    `gorm:"type:char(36);primaryKey" json:"id"`
	UserIDA   string    `gorm:"type:char(36);not null;uniqueIndex:uk_friends;index:idx_user_a" json:"user_id_a"`
	UserIDB   string    `gorm:"type:char(36);not null;uniqueIndex:uk_friends;index:idx_user_b" json:"user_id_b"`
	Status    string    `gorm:"type:varchar(20);default:'pending'" json:"status"`
	CreatedAt time.Time `gorm:"type:datetime(3)" json:"created_at"`
	UpdatedAt time.Time `gorm:"type:datetime(3)" json:"updated_at"`

	UserA User `gorm:"foreignKey:UserIDA" json:"user_a,omitempty"`
	UserB User `gorm:"foreignKey:UserIDB" json:"user_b,omitempty"`
}

func (Friendship) TableName() string {
	return "friendships"
}

// Leaderboard 排行榜快照模型
type Leaderboard struct {
	ID         string    `gorm:"type:char(36);primaryKey" json:"id"`
	RouteID    string    `gorm:"type:char(36);not null;uniqueIndex:uk_route_user;index:idx_route_time" json:"route_id"`
	UserID     string    `gorm:"type:char(36);not null;uniqueIndex:uk_route_user;index:idx_user" json:"user_id"`
	RunID      string    `gorm:"type:char(36);not null" json:"run_id"`
	TotalTime  int64     `gorm:"type:int;not null" json:"total_time"`
	AvgPace    *int      `gorm:"type:int" json:"avg_pace"`
	RunCount   int       `gorm:"type:int;default:0" json:"run_count"`
	RecordedAt time.Time `gorm:"type:datetime(3);not null" json:"recorded_at"`
	CreatedAt  time.Time `gorm:"type:datetime(3)" json:"created_at"`
	UpdatedAt  time.Time `gorm:"type:datetime(3)" json:"updated_at"`

	Route Route `gorm:"foreignKey:RouteID" json:"route,omitempty"`
	User  User  `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Run   Run   `gorm:"foreignKey:RunID" json:"run,omitempty"`
}

func (Leaderboard) TableName() string {
	return "route_leaderboards"
}

File: stride_moor_app/lib/core/models/challenge.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'challenge.freezed.dart';
part 'challenge.g.dart';

/// 路线挑战（陪跑邀请）
///
/// 【注意】手写 fromJson 以正确映射后端 snake_case 字段
@freezed
class Challenge with _$Challenge {
  const factory Challenge({
    required String id,
    required String routeId,
    /// 挑战者
    required String challengerId,
    String? challengerRunId,
    /// 被邀请者
    String? inviteeId,
    /// 陪跑模式: real_replay / constant / rabbit / tortoise_hare / goal
    @Default('real_replay') String ghostMode,
    /// 目标挑战维度: pace / heart_rate / cadence / stride_length
    String? goalMetric,
    /// 状态: pending / running / completed
    @Default('pending') String status,
    DateTime? createdAt,
    DateTime? completedAt,
  }) = _Challenge;

  /// 手写 fromJson —— 后端返回 snake_case
  factory Challenge.fromJson(Map<String, dynamic> json) {
    return Challenge(
      id: json['id'] as String? ?? '',
      routeId: json['route_id'] as String? ?? '',
      challengerId: json['challenger_id'] as String? ?? '',
      challengerRunId: json['challenger_run_id'] as String?,
      inviteeId: json['invitee_id'] as String?,
      ghostMode: json['ghost_mode'] as String? ?? 'real_replay',
      goalMetric: json['goal_metric'] as String?,
      status: json['status'] as String? ?? 'pending',
      createdAt: json['created_at'] == null
          ? null
          : DateTime.parse(json['created_at'] as String),
      completedAt: json['completed_at'] == null
          ? null
          : DateTime.parse(json['completed_at'] as String),
    );
  }
}

/// 对比报告
///
/// 【注意】手写 fromJson 以正确映射后端 snake_case 字段
@freezed
class Comparison with _$Comparison {
  const factory Comparison({
    required String id,
    String? challengeId,
    required String runAId,
    required String runBId,
    /// 总体差异摘要
    Map<String, dynamic>? overallDiff,
    /// 分段对比详情 JSON
    @Default({}) Map<String, dynamic> splitsJson,
    /// AI诊断建议 JSON
    @Default({}) Map<String, dynamic> diagnosisJson,
    DateTime? createdAt,
  }) = _Comparison;

  /// 手写 fromJson —— 后端返回 snake_case
  factory Comparison.fromJson(Map<String, dynamic> json) {
    return Comparison(
      id: json['id'] as String? ?? '',
      challengeId: json['challenge_id'] as String?,
      runAId: json['run_a_id'] as String? ?? '',
      runBId: json['run_b_id'] as String? ?? '',
      overallDiff: json['overall_diff'] as Map<String, dynamic>?,
      splitsJson: json['splits_json'] as Map<String, dynamic>? ?? const {},
      diagnosisJson:
          json['diagnosis_json'] as Map<String, dynamic>? ?? const {},
      createdAt: json['created_at'] == null
          ? null
          : DateTime.parse(json['created_at'] as String),
    );
  }
}
